hive-metastore/src/main/java/org/apache/iceberg/hive/HiveCatalogs.java

```diff
@@ -36,4 +38,11 @@ public static HiveCatalog loadCatalog(Configuration conf) {
     String metastoreUri = conf.get(HiveConf.ConfVars.METASTOREURIS.varname, "");
     return CATALOG_CACHE.get(metastoreUri, uri -> new HiveCatalog(conf));
   }
+
+  public static HiveCatalog loadCatalog(String catalogName, Map<String, String> properties, Configuration conf) {
+    // metastore URI can be null in local mode
+    String metastoreUri = conf.get(HiveConf.ConfVars.METASTOREURIS.varname, "");
+    return CATALOG_CACHE.get(metastoreUri, uri -> (HiveCatalog) CatalogUtil.loadCatalog(HiveCatalog.class.getName(),
+            catalogName, properties, conf));
+  }
```

Comment:
With the new Catalog system I think the key in the cache could be the catalogName since we expect it to be unique. I am not sure what to do with the original default behavior. Maybe the `default` name could be ok there as well.

Reply:
I like the idea of having the catalogName as key. It's much cleaner than having the metastore uri and it's easier to track back. 
The loadCatalog(conf) method was only used from the TestTables, so I extended it with an additional catalogName param.

Reply:
FWIW, This class has been dropped in other places in favour of `CustomCatalogs` or other catalog-name based caching mechanism. So rather than cache just the Hive catalog all catalogs are cached by catalog name

Reply:
I wonder if caching logic could be added to `Catalogs` and this class could be marked deprecated

Reply:
I understand your point, but to be honest I would not introduce any caching related changes to this PR. I could do the refactoring in a separate PR. What do you say?

Reply:
I would prefer it in this PR but don't feel that strongly about it. My reasoning is that technically there are already caching related changes: changing the cache key in `HiveCatalogs` and not caching any of the custom catalogs. To me it makes sense cache all catalogs in 1 place.

Reply:
On a side note changing the `HiveCatalogs` to cache only by catalog name could mean two users with different settings/permissions but same catalog name could potentially resolve the same HiveCatalog in the cache, which may be wrong. A per session or per user cache (similar to `CustomCatalogs` in spark2) rather than a global static cache may be safer. However I am far from a Hive expert so I don't know how achievable that is.

Reply:
I agree with @rymurr here. The `HiveCatalogs` class was used by Spark when the catalog was implicit and not a configured catalog. Now, we've moved to create a Spark catalog for that case that is named `default_iceberg`.

I don't think it makes sense to re-use this class for catalogs that are configured and named. Those should be loaded and cached separately. I would probably also mirror the changes in Spark 2's `IcebergSource` that Ryan made and follow his advice to deprecate and remove this class.

Reply:
was the plan still to not change this and just add a deprecated ?

Reply:
I agree with @rymurr's question. I thought that this PR hadn't been updated yet because this method is still here.

Do you plan to use it? I don't see how it fits in.

Reply:
The plan is to remove the `HiveCatalogs` completely but in a separate PR. I would like to keep it for time being, because of it's caching mechanism. Without it, we would need to create a new Hive client every time the `loadTable()`, `createTable()` or `dropTable()` is called, which is really expansive.

In a different PR, I will work on introducing a more general cache (not just for Hive catalogs), similar to `CustomCatalogs` cache.

Reply:
I'll take a look at what you have here, but it seems like a problem that needs to be solved in this PR. HiveCatalogs can only solve that problem for the default Hive catalog, not for other catalogs that this PR adds support for. I think that caching for alternative catalogs is just as important for the same reasons.

Reply:
> I'll take a look at what you have here, but it seems like a problem that needs to be solved in this PR. HiveCatalogs can only solve that problem for the default Hive catalog, not for other catalogs that this PR adds support for. I think that caching for alternative catalogs is just as important for the same reasons.

If we would cache the Catalogs by name+session as it is done in Spark, then it would mean 4 extra HMS connections (2 HS2, 2 Tez/AM) per user. On bigger clusters this could cause issues.

Also since Hive users can change the configuration on session level, caching becomes even more complicated, since old cached Catalog might have a different configuration than the one the user actually requests.

One solution could be to require the catalog implementation to provide its own key for caching based on the configuration. Adding this to this PR might increase the size of it seriously.

Do I overcomplicate things? Do you see a simplier solution?

Reply:
The concern about the number of connections to HMS makes sense, but we should check some assumptions:
1. Are you saying that there will be 2 on HS2 and 2 on the AM because the pool size is 2? If there aren't concurrent uses, then the pool won't actually grow to the maximum size. Maybe the AM only has 1?
2. Is more than one connection abnormal? If so, Hive could set the pool size to 1.
3. Does table serialization help here to avoid using the pool on the AM?
4. Can we share client pools across catalogs?

I"d rather take care of this problem using stricter settings.

Reply:
> 1. Are you saying that there will be 2 on HS2 and 2 on the AM because the pool size is 2? If there aren't concurrent uses, then the pool won't actually grow to the maximum size. Maybe the AM only has 1?
> 2. Is more than one connection abnormal? If so, Hive could set the pool size to 1.
> 3. Does table serialization help here to avoid using the pool on the AM?

Theoretically there can be concurrent uses for the Catalogs on HS2 side, but you are right that this will be a single connection in most cases. And the `ClientPool` calls are serialized by the `run` method if the pool is exhausted, so it could be ok in a sense that the queries could work but the number of HMS connections per session is still doubled.

For writes TezAM (for Tez) or YARN Application Master (for MR) calls the commitJob which still needs to create the Catalog. @marton-bod will look into this how can we push the commit to the HS2 side.

> 4. Can we share client pools across catalogs?

Currently under the hood we do something like this (share the Catalog and with that, the ClientPool too 😄)

OTOH, when connecting to HMS, clients can set their own settings:
- One partially relevant example is #2070
- More relevant is that connection timeout, directSql settings and other stuff can be changed as well
- In theory the users connecting to the HMS could user their own credentials (some deployments are using a service account)

So I think sharing HMS Clients between the Catalogs is not really a good practice.

I think the best solution would be a different HiveCatalog implementation reusing the HMS Client from the Hive session. This is guaranteed to have the correct settings. It should be noted that the HMS client instance should be refreshed before every call, since Hive can close the client if settings are changed. Also it strengthens the dependency on Hive which might not be desirable.

All that said, I think the `name` based caching is flawed if the configuration can be changed (as it could be happen in case of Hive ATM).
Imagine this situation:
- Expert user creates a new catalog by setting the configs.
- Queries a table and founds out that one of the configs is not correctly set (wrong URI or warehouse root)
- Fixes the wrong config
- Queries the table again. Since the name did not change the old Catalog is returned, so everything is failing again.

Actually @lcspinter found the same issue in some of the unit tests 😄

Reply:
> For writes TezAM (for Tez) or YARN Application Master (for MR) calls the commitJob which still needs to create the Catalog.

This is later in the job, though. I wonder if we should introduce a thread to close connections if they are not used for some period of time. Garbage collecting connections could help if we think that the extra connection from the HS2 side is a problem. The AM connection probably wouldn't last long though?

> I think sharing HMS Clients between the Catalogs is not really a good practice.

Sounds reasonable.

> I think the best solution would be a different HiveCatalog implementation reusing the HMS Client from the Hive session.

This sounds okay, but should probably be done in a follow-up instead of here. It also doesn't affect name-based caching. The only catalog that could share a connection with the Hive session is the unnamed catalog used when there is no catalog set. We cannot share connections to named catalogs because there is no reason to think those are hitting the same metastore.

> ll that said, I think the name based caching is flawed if the configuration can be changed

I don't think that configuration can be changed, for the reason you cite. Once a catalog is created, it is configured and that can't be changed. We _could_ look up settings each time they are used but that seems like a lot of work for a problem we fundamentally don't need to solve.

Reply:
> Garbage collecting connections could help if we think that the extra connection from the HS2 side is a problem. The AM connection probably wouldn't last long though?

Garbage collection could help with inactive sessions, but connecting the catalog to the session and removing it on session close could be a better solution (if possible somehow).

About the AM connection - I am not sure how it works with YARN, but with TezAM the Catalog/client should live until the session is there. My hope that we can get rid of this altogether in the near future.

My main concern is that even with the best efforts we basically double the HMS connections if all of the users start to use Iceberg tables.

> > ll that said, I think the name based caching is flawed if the configuration can be changed
> 
> I don't think that configuration can be changed, for the reason you cite. Once a catalog is created, it is configured and that can't be changed. We _could_ look up settings each time they are used but that seems like a lot of work for a problem we fundamentally don't need to solve.

The thing I love about Iceberg is that any data scientist can create an S3 bucket and put data there with the corresponding metadata and then can publish it. Another data scientist can just get the location of the table, and easily can create a table reading that data and then can aggregate these from multiple sources.
These data scientists are quite capable guys and can configure their own catalogs, and they do not want to wait for an administrator to configure the catalog for them. So they would need the ability to configure the catalog themselves. On the other hand if they are anything like me then they are quite capable of making mistakes too 😄 and they will create misconfigured catalogs and would like to fix those without closing the Hive session.

After some more thoughts I think the best solution would be to add a general parameter to the catalogs, like `cacheable` which defaults to true, but if it is false then the the catalog should skip/remove the old version with the same name from the cache and create a new one.

This could fix the problem with the name based caching, and the "only" remaining one is the doubling of the HMS connection.

Reply:
I guess another option is to detect changes and refresh the cached catalog if there are changes. That seems a bit expensive for every catalog lookup.